Exception during zip operations: de.schlichtherle.truezip.io.InputException: java.util.zip.ZipException: ??? (expected Local File Header)

**Problem**: When trying to build a docker image, the following exception occurs (full stacktrace at alf1 link)

> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':apix-docker:docker-61:applyAlfrescoAmp'.
> > 11100015 An error was encountered during deployment of the AMP into the WAR: 11100014 Unable to load module details from property file. File Not Found, /home/red/Documents/code/xenit/alfred-api/apix-impl/61/build/dist/apix-impl-61-2.3.0-local-SNAPSHOT.amp/module.properties
> 
> * Try:
> Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
> 
> * Get more help at https://help.gradle.org
> 
> Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
> Use '--warning-mode all' to show the individual deprecation warnings.
> See https://docs.gradle.org/5.6/userguide/command_line_interface.html#sec:command_line_warnings
> 
> BUILD FAILED in 2s
> 20 actionable tasks: 8 executed, 12 up-to-date
> Cause: module.properties (expected Local File Header)
> 14:03:35: Task execution finished 'buildDockerImage'.

**Observations**
* The amp was build with the gradle-sdk v1.0.0
* The amp artifact was referenced as another gradle project int he build.gradle file
* The amp artifact does contain a correct module.properties file
* The problem occurs intermittently, but with unknown cause or solution
* The problem only occurs with alfresco 6.x builds

**resources**
Alf1 folder containing full error log and 2 faulty artifacts: [link](https://alf1.xenit.eu/share/page/folder-details?nodeRef=workspace://SpacesStore/4f479f65-1558-41d1-bedf-f9adf9ead236)

--- a/src/integrationTest/java/eu/xenit/gradle/testrunner/AbstractIntegrationTest.java
+++ b/src/integrationTest/java/eu/xenit/gradle/testrunner/AbstractIntegrationTest.java
@@ -92,7 +92,6 @@ public abstract class AbstractIntegrationTest {
                 .withArguments(task, "--stacktrace", "--rerun-tasks", "--info")
                 .withGradleVersion(gradleVersion)
                 .withPluginClasspath()
-                .withDebug(true)
                 .forwardOutput();
     }
 

--- a/src/main/java/eu/xenit/gradle/tasks/InstallAmpsInWarTask.java
+++ b/src/main/java/eu/xenit/gradle/tasks/InstallAmpsInWarTask.java
@@ -35,18 +35,20 @@ public class InstallAmpsInWarTask extends InjectFilesInWarTask {
 
     @Override
     public void injectFiles() throws IOException {
-        ModuleManagementTool moduleManagementTool = new ModuleManagementTool();
         FileUtils.copyFile(getInputWar(), getOutputWar());
+        Util.withGlobalTvfsLock(() -> {
+            ModuleManagementTool moduleManagementTool = new ModuleManagementTool();
+            Set<File> sourceFiles = getSourceFiles().stream()
+                    .flatMap(InstallAmpsInWarTask::listFilesRecursively)
+                    .collect(Collectors.toSet());
+            List<File> filesInInstallationOrder = ModuleDependencySorter.sortByInstallOrder(sourceFiles, getOutputWar());
 
-        Set<File> sourceFiles = getSourceFiles().stream()
-                .flatMap(InstallAmpsInWarTask::listFilesRecursively)
-                .collect(Collectors.toSet());
-        List<File> filesInInstallationOrder = ModuleDependencySorter.sortByInstallOrder(sourceFiles, getOutputWar());
+            for (File file : filesInInstallationOrder) {
+                getLogger().debug("installing amp from {} in war {}",file.getAbsolutePath(), getOutputWar().getAbsolutePath());
+                moduleManagementTool
+                        .installModule(file.getAbsolutePath(), getOutputWar().getAbsolutePath(), false, true, false);
+            }
 
-        for (File file : filesInInstallationOrder) {
-            getLogger().debug("installing amp from {} in war {}",file.getAbsolutePath(), getOutputWar().getAbsolutePath());
-            moduleManagementTool
-                    .installModule(file.getAbsolutePath(), getOutputWar().getAbsolutePath(), false, true, false);
-        }
+        });
     }
 }

--- a/src/main/java/eu/xenit/gradle/tasks/Util.java
+++ b/src/main/java/eu/xenit/gradle/tasks/Util.java
@@ -13,22 +13,33 @@ import java.util.function.Consumer;
 
 class Util {
 
+    private static final Object TVFS_LOCK = new Object();
+
+    static void withGlobalTvfsLock(Runnable runnable) {
+        Objects.requireNonNull(runnable);
+        synchronized (TVFS_LOCK) {
+            runnable.run();
+        }
+    }
+
     static void withWar(File warFile, Consumer<TFile> closure) {
         Objects.requireNonNull(warFile, "warFile");
         Objects.requireNonNull(closure, "closure");
+        withGlobalTvfsLock(() -> {
             TConfig config = TConfig.get();
             config.setArchiveDetector(new TArchiveDetector("war|amp", new JarDriver(IOPoolLocator.SINGLETON)));
             TFile archive = new TFile(warFile);
-        try {
-            closure.accept(archive);
-        } finally {
             try {
-                TVFS.umount(archive);
-            } catch (FsSyncException ignored) {
-                // This exception is intentionally ignored.
-                // Throwing exceptions in a finally block would swallow the original exception
-                // And it is no big deal if the archive can not be unmounted now, it will be unmounted during process shutdown anyways
+                closure.accept(archive);
+            } finally {
+                try {
+                    TVFS.umount(archive);
+                } catch (FsSyncException ignored) {
+                    // This exception is intentionally ignored.
+                    // Throwing exceptions in a finally block would swallow the original exception
+                    // And it is no big deal if the archive can not be unmounted now, it will be unmounted during process shutdown anyways
+                }
             }
-        }
+        });
     }
 }

